vowpalwabbit/offset_tree.cc

```diff
@@ -0,0 +1,207 @@
+    // deal with degenerate cases of 0 and 1 actions
+    if (_num_leaf_nodes == 0)
+    {
+      _initialized = true;
+      return;
+    }
+
+    try
+    {
+      nodes.reserve(2 * _num_leaf_nodes - 1);
+      std::vector<uint32_t> tournaments;
+      // Add all leaf nodes to node collection
+      tournaments.reserve(_num_leaf_nodes);
+      for (uint32_t i = 0; i < _num_leaf_nodes; i++)
+      {
+        nodes.emplace_back(i, 0, 0, true);
+        tournaments.emplace_back(i);
+      }
+      while (tournaments.size() > 1)
+      {
+        // ids of next set of nodes start with the current
+        const auto num_tournaments = tournaments.size();
+        // node count (since ids are zero based)
+        auto id = nodes.size();
+        std::vector<uint32_t> new_tournaments;
+        new_tournaments.reserve(num_tournaments / 2 + 1);
+        for (size_t j = 0; j < num_tournaments / 2; ++j)
+        {
+          auto left = tournaments[2 * j];
+          auto right = tournaments[2 * j + 1];
+          new_tournaments.emplace_back(id);
+          nodes.emplace_back(id++, left, right, false);
+        }
+        if (num_tournaments % 2 == 1)
+          new_tournaments.emplace_back(tournaments.back());
+        tournaments = std::move(new_tournaments);
+      }
+      root_idx = tournaments[0];
+    }
+    catch (std::bad_alloc& e)
+    {
+      THROW("Unable to allocate memory for offset tree.  Label count:" << _num_leaf_nodes << " bad_alloc:" << e.what());
+    }
+    _initialized = true;
+  }
+
+  uint32_t min_depth_binary_tree::internal_node_count() const { return nodes.size() - _num_leaf_nodes; }
+
+  uint32_t min_depth_binary_tree::leaf_node_count() const { return _num_leaf_nodes; }
+
+  void offset_tree::init(uint32_t num_actions) { binary_tree.build_tree(num_actions); }
+
+  int32_t offset_tree::learner_count() const { return binary_tree.internal_node_count(); }
+
+  const std::vector<float>& offset_tree::predict(LEARNER::single_learner& base, example& ec)
+  {
+    // - pair<float,float> stores the scores for left and right nodes
+    // - prediction_buffer stores predictions for all the nodes in the tree for the duration
+    //   of the predict() call
+    // - static thread_local ensures only one copy per calling thread.  This is to reduce
+    //   memory allocations in steady state.
+    static thread_local std::vector<std::pair<float, float>> prediction_buffer(learner_count());
+    static thread_local std::vector<float> scores(binary_tree.leaf_node_count());
+
+    auto& t = binary_tree;
+
+    prediction_buffer.clear();
+    scores.resize(t.leaf_node_count());
+
+    // Handle degenerate cases of zero and one node trees
+    if (t.leaf_node_count() == 0)
+      return scores;
+    if (t.leaf_node_count() == 1)
+    {
+      scores[0] = 1.0f;
+      return scores;
+    }
+
+    // Get predictions for all internal nodes
+    for (auto idx = 0; idx < t.internal_node_count(); ++idx)
+    {
+      base.predict(ec, idx);
```

Comment:
There is a potential optimization here in that we could use a multi_predict operation.   Probably, we should save that for later.

Reply:
I added an enhancement request to track it.  Unclear to me what multi_predict does.  Do you mind elaborating?